README
Gather random book data by scraping Goodreads.

Thanks to Code Review user MrGrj for the suggested improvements (present in v2
onward) and user alecxe for the suggested code improvements (present in v5).


CHANGELOG
v5 (current):

 - Added code to maintain session for URL
 - Changed parser from html.parser to lxml
 - Added SoupStrainer to parse more efficiently
 - Changed single line comments to use '#' instead of docstrings
 - Removed random endline spaces, added spaces where they should go
 - Changed comma to colon for soup.find() in get_book_rating_count
 - Updated command line parsing help print out

v4:

 - Added command line parsing
 - In to_int(rating_count) added if-else statement to check if rating count is
   already an integer
 - Fixed issue caused by the removal of series identification where an extra
   space is left at the end of the title
 - Added option to continually print total line count after appendage (always
   prints on the same line) if --verbosity is 1 or 3
 - Added option to print program run time (in real time) and total line count of
   lines added upon keyboard interrupt if --verbosity is 2 or 3
 - Added FileNotFoundError to main try-except block to catch if file does not
   exist
 - Now actually prints something if ConnectionResetError is caught

v3:

 - Program now outputs to a file instead of printing
 - Automatically creates data.csv in working directory
 - Outputs with CSV formatting (double quotes around title, comma separation,
   etc.)

v2:

 - Error catching (not all, but some)
 - Now using BeautifulSoup instead of regex

v1:

 - Prints out data. Continues until it is manually closed or crashes because of
   an error.


Booky.py
import requests
import re

URL = "https://www.goodreads.com/book/random"

while True:
	html_text = requests.get(URL).text

	#	Rating Count
	bg_rating_count 	= html_text.find("<span class=\"value-title\" title=\"") + 33
	end_rating_count	= html_text.find("\" itemprop=\"ratingCount\">", bg_rating_count)
	rating_count 		= int(html_text[bg_rating_count : end_rating_count].replace(',', ''))

	if rating_count >= 30:
		if "<span itemprop=\"numberOfPages\">" in html_text:
			
			#	Title
			bg_title	= html_text.find("<meta property=\"og:title\" content=\"") + 35
			end_title	= html_text.find("\"/>", bg_title)
			title = html_text[bg_title : end_title].replace("&amp;", '&')
			title = re.sub(r'\((.*)\)','', title)

			#	Pages
			bg_pages	= html_text.find("<span itemprop=\"numberOfPages\">") + 31
			end_pages	= html_text.find(" page", bg_pages)
			pages 		= int(html_text[bg_pages : end_pages])

			#	Rating
			bg_rating	= html_text.find("<span class=\"average\" itemprop=\"ratingValue\">") + 45
			end_rating	= html_text.find("</", bg_rating)
			rating		= float(html_text[bg_rating : end_rating])

			if pages != 0:
				print(title, pages, rating, sep='\t')


Booky-v1.py
import requests
import re

URL = "https://www.goodreads.com/book/random"

while True:
	html_text = requests.get(URL).text

	# Rating Count
	bg_rating_count 	= html_text.find("<span class=\"value-title\" title=\"") + 33
	end_rating_count	= html_text.find("\" itemprop=\"ratingCount\">", bg_rating_count)
	rating_count 		= int(html_text[bg_rating_count : end_rating_count].replace(',', ''))

	if rating_count >= 30:
		if "<span itemprop=\"numberOfPages\">" in html_text:
			
			# Title
			bg_title	= html_text.find("<meta property=\"og:title\" content=\"") + 35
			end_title	= html_text.find("\"/>", bg_title)
			title = html_text[bg_title : end_title].replace("&amp;", '&')
			title = re.sub(r'\((.*)\)','', title)

			# Pages
			bg_pages	= html_text.find("<span itemprop=\"numberOfPages\">") + 31
			end_pages	= html_text.find(" page", bg_pages)
			pages 		= int(html_text[bg_pages : end_pages])

			# Rating
			bg_rating	= html_text.find("<span class=\"average\" itemprop=\"ratingValue\">") + 45
			end_rating	= html_text.find("</", bg_rating)
			rating		= float(html_text[bg_rating : end_rating])

			if pages != 0:
				print(title, pages, rating, sep='\t')


Booky-v2.py
from bs4 import BeautifulSoup as bs
import requests
import re

URL = "https://www.goodreads.com/book/random"
MIN_RATING_COUNT = 30


def to_int(rating_count):
	""" Return rating count as an int """
	rating_count = rating_count.split()[0]
	if ',' in rating_count:
		return int(rating_count.replace(',', ''))
	return int(rating_count)


def format_title(book_title):
	""" 
		Return formatted book title:
			"Actual Book Title! (Series identifier #3)" -> "Actual Book Title!"
			"History of 4D Fish Slapping [Silly, Part 4]" -> "History of 4D Fish Slapping"
	"""
	book_title = ' '.join(book_title.split()).replace('&amp;', '&')
	book_title = re.sub(r'(\(|\[)(.*)(\)|\])','', book_title)
	return book_title


def get_html_source():
	""" Return html source """
	html_source = requests.get(URL).text
	return bs(html_source, 'html.parser')


def get_book_rating_count(soup):
	""" Return book rating count """
	return soup.find('span', attrs={'class', 'value-title'}).get_text()


def get_book_title(soup):
	""" Return book title"""
	return soup.find('h1', attrs={'class': 'bookTitle'}).get_text()


def get_book_pages(soup):
	""" Return book pages """
	book_pages = soup.find('span', attrs={'itemprop': 'numberOfPages'}).get_text()
	if " pages" in book_pages:
		return book_pages.replace(' pages', '')
	elif " page" in book_pages:
		return book_pages.replace(' page', '')
	return book_pages


def get_book_rating(soup):
	""" Return book rating """
	return soup.find('span', attrs={'itemprop': 'ratingValue'}).get_text()


def main():
	""" Gather book data until the program is manually closed or connection issue. """

	while True:
		soup = get_html_source()
		book_rating_count = get_book_rating_count(soup)

		if to_int(book_rating_count) >= MIN_RATING_COUNT:
			try:
				book_pages = get_book_pages(soup)
				book_title = format_title(get_book_title(soup))
				book_rating = get_book_rating(soup)
				if book_pages != "0":
					print(book_title, book_pages, book_rating, sep = '\t')
			except AttributeError:
				continue

if __name__ == '__main__':
	try:
		main()
	except KeyboardInterrupt:
		""" Manually close program """ 
		# print("Exiting program...")
		pass
	except ConnectionResetError:
		""" Connection error, most commonly caused by connection reset by peer """
		# print("Connection reset by peer...")
		pass


Booky-v3.py
from bs4 import BeautifulSoup as bs
import requests
import re

URL = "https://www.goodreads.com/book/random"
MIN_RATING_COUNT = 30
FILE_PATH = "data.csv"


def to_int(rating_count):
	""" Return rating count as an int """
	rating_count = rating_count.split()[0]
	if ',' in rating_count:
		return int(rating_count.replace(',', ''))
	return int(rating_count)


def format_title(book_title):
	""" 
		Return formatted book title:
			"Actual Book Title! (Series identifier #3)" -> "Actual Book Title!"
			"History of 4D Fish Slapping [Silly, Part 4]" -> "History of 4D Fish Slapping"

		Along with CSV formatting
	"""
	book_title = ' '.join(book_title.split()).replace('&amp;', '&')
	book_title = re.sub(r'(\(|\[)(.*)(\)|\])','', book_title)
	book_title = '"' + book_title.replace('"', '""') + '"'
	return book_title


def get_html_source():
	""" Return html source """
	html_source = requests.get(URL).text
	return bs(html_source, 'html.parser')


def get_book_rating_count(soup):
	""" Return book rating count """
	return soup.find('span', attrs={'class', 'value-title'}).get_text()


def get_book_title(soup):
	""" Return book title"""
	return soup.find('h1', attrs={'class': 'bookTitle'}).get_text()


def get_book_pages(soup):
	""" Return book pages """
	book_pages = soup.find('span', attrs={'itemprop': 'numberOfPages'}).get_text()
	if " pages" in book_pages:
		return book_pages.replace(' pages', '')
	elif " page" in book_pages:
		return book_pages.replace(' page', '')
	return book_pages


def get_book_rating(soup):
	""" Return book rating """
	return soup.find('span', attrs={'itemprop': 'ratingValue'}).get_text()


def main():
	""" Gather book data until the program is manually closed or connection issue. """

	while True:
		soup = get_html_source()
		book_rating_count = get_book_rating_count(soup)

		if to_int(book_rating_count) >= MIN_RATING_COUNT:
			try:
				book_pages = get_book_pages(soup)
				book_title = format_title(get_book_title(soup))
				book_rating = get_book_rating(soup)
				if book_pages != "0":
					with open(FILE_PATH, 'a') as file:
						file.write(book_title + ','  + book_pages + ',' + book_rating + '\n')

			except AttributeError:
				continue

if __name__ == '__main__':
	try:
		main()
	except KeyboardInterrupt:
		""" Manually close program """ 
		# print("Exiting program...")
		pass
	except ConnectionResetError:
		""" Connection error, most commonly caused by connection reset by peer """
		# print("Connection reset by peer...")
		pass


Booky-v5.py
import argparse
from bs4 import BeautifulSoup as bs
from bs4 import SoupStrainer
import os.path
import requests
import re
import sys
import time

URL = "https://www.goodreads.com/book/random"


def to_int(rating_count):
	# Return rating count as an int
	if isinstance(rating_count, int):
		return rating_count
	else:
		rating_count = rating_count.split()[0]
		if ',' in rating_count:
			return int(rating_count.replace(',', ''))
		return int(rating_count)


def format_title(book_title):
	"""
		Return formatted book title:
			"Actual Book Title! (Series identifier #3)" -> "Actual Book Title!"
			"History of 4D Fish Slapping [Silly, P4]" -> "History of 4D Fish Slapping"

		Along with CSV formatting
	"""
	book_title = ' '.join(book_title.split()).replace('&amp;', '&')
	book_title = re.sub(r'(\(|\[)(.*)(\)|\])', '', book_title)
	book_title = '"' + book_title.replace('"', '""') + '"'
	if book_title[-2:-1] == ' ':
		book_title = book_title[:-2] + '"'
	return book_title


def get_html_source(session):
	# Return html source
	html_source = session.get(URL).text
	return bs(html_source, 'lxml', parse_only=SoupStrainer(id="metacol"))


def get_book_rating_count(soup):
	# Return book rating count
	try:
		book_rating_count = soup.find('span', attrs={'class': 'value-title'}).get_text()
		return book_rating_count
	except AttributeError:
		# Attribute error, the rating count is missing
		return -1


def get_book_title(soup):
	# Return book title
	return soup.find('h1', attrs={'class': 'bookTitle'}).get_text()


def get_book_pages(soup):
	# Return book pages
	book_pages = soup.find('span', attrs={'itemprop': 'numberOfPages'}).get_text()
	if " pages" in book_pages:
		return book_pages.replace(' pages', '')
	elif " page" in book_pages:
		return book_pages.replace(' page', '')
	return book_pages


def get_book_rating(soup):
	# Return book rating
	return soup.find('span', attrs={'itemprop': 'ratingValue'}).get_text()


def main():
	# Command line parsing: verbosity, min rating, and required file path
	parser = argparse.ArgumentParser(description="Gather random book data from Goodreads and append it to a file in CSV format, \
until the program is manually closed or until a connection issue. I recommend having the CSV header: Title,Pages,Rating.",
		formatter_class=argparse.RawTextHelpFormatter)
	parser.add_argument("-v", "--verbosity", type=int, choices=[0, 1, 2, 3],
		help="""specify verbosity level, default = 2
0 = no output
1 = continually print line count after appending to file
2 = after keyboard interrupt, print count of lines added since program start and program run time
3 = both verbosity options 1 and 2""", default=2)
	parser.add_argument("-mr", "--min-rating", type=int, default=30,
		help="specify the minimum accepted rating, default = 30")
	parser.add_argument("filepath", help="output file, only supports FULL path (no tilde, etc.)")
	args = parser.parse_args()

	FILE_PATH = args.filepath
	MIN_RATING_COUNT = args.min_rating

	try:
		if args.verbosity in (2, 3):
			start_time = time.time()
			with open(FILE_PATH, 'r') as file:
				init_filelength = sum(1 for _ in file)

		s = requests.Session()

		while True:
			soup = get_html_source(s)
			book_rating_count = get_book_rating_count(soup)

			if to_int(book_rating_count) >= MIN_RATING_COUNT:
				try:
					book_pages = get_book_pages(soup)
					book_title = format_title(get_book_title(soup))
					book_rating = get_book_rating(soup)
					# Re-check if file exists to catch deletion while program is running
					if book_pages != "0" and os.path.isfile(FILE_PATH):
						with open(FILE_PATH, 'a') as file:
							file.write(book_title + ',' + book_pages + ',' + book_rating + '\n')

						if args.verbosity in (1, 3):
							sys.stdout.write("\033[F")
							with open(FILE_PATH, 'r') as file:
								print("\nLine count:", sum(1 for _ in file), end='')
					elif not os.path.isfile(FILE_PATH):
						raise FileNotFoundError

				except AttributeError:
					continue

	except FileNotFoundError:
		print("File '", FILE_PATH, "' does not exist, or has been deleted.", sep='')

	except KeyboardInterrupt:
		if args.verbosity in (2, 3):
			with open(FILE_PATH, 'r') as file:
				lines_added = sum(1 for _ in file) - init_filelength
			print("\nLines added: ", lines_added, " (since program start)", sep='')
			print("Run time: ", round(time.time() - start_time, 3), " seconds (real)", sep='')

if __name__ == '__main__':
	try:
		main()

	except ConnectionResetError:
		print("\nConnection reset by peer, exiting program.")
		pass


Booky-v4.py
import argparse
from bs4 import BeautifulSoup as bs
import os.path
import requests
import re
import sys
import time

URL = "https://www.goodreads.com/book/random"


def to_int(rating_count):
	""" Return rating count as an int """
	if isinstance(rating_count, int):
		return rating_count
	else:
		rating_count = rating_count.split()[0]
		if ',' in rating_count:
			return int(rating_count.replace(',', ''))
		return int(rating_count)


def format_title(book_title):
	""" 
		Return formatted book title:
			"Actual Book Title! (Series identifier #3)" -> "Actual Book Title!"
			"History of 4D Fish Slapping [Silly, Part 4]" -> "History of 4D Fish Slapping"

		Along with CSV formatting
	"""
	book_title = ' '.join(book_title.split()).replace('&amp;', '&')
	book_title = re.sub(r'(\(|\[)(.*)(\)|\])','', book_title)
	book_title = '"' + book_title.replace('"', '""') + '"'
	if book_title[-2:-1] == ' ':
		book_title = book_title[:-2] + '"'
	return book_title


def get_html_source():
	""" Return html source """
	html_source = requests.get(URL).text
	return bs(html_source, 'html.parser')


def get_book_rating_count(soup):
	""" Return book rating count """
	try:
		book_rating_count = soup.find('span', attrs={'class', 'value-title'}).get_text()
		return book_rating_count
	except AttributeError:
		""" Attribute error, the rating count is missing """
		return -1


def get_book_title(soup):
	""" Return book title"""
	return soup.find('h1', attrs={'class': 'bookTitle'}).get_text()


def get_book_pages(soup):
	""" Return book pages """
	book_pages = soup.find('span', attrs={'itemprop': 'numberOfPages'}).get_text()
	if " pages" in book_pages:
		return book_pages.replace(' pages', '')
	elif " page" in book_pages:
		return book_pages.replace(' page', '')
	return book_pages


def get_book_rating(soup):
	""" Return book rating """
	return soup.find('span', attrs={'itemprop': 'ratingValue'}).get_text()


def main():
	""" Command line parsing, with specifiable verbosity, minimum rating, and required file name """
	parser = argparse.ArgumentParser(description="Gather random book data from Goodreads and append it to a file in CSV format, until the program is manually closed or until a connection issue.\
		\n\nexample command: Booky-v4.py -v3 -mr /path/to/file.csv",
		formatter_class=argparse.RawTextHelpFormatter)
	parser.add_argument("-v", "--verbosity", type=int, choices=[0,1,2,3],
		help="specify verbosity level, default = 2\
		\n0 = no output\
		\n1 = continually print line count after appending to file\
		\n2 = after keyboard interrupt, print count of lines added since program start and program run time\
		\n3 = both verbosity options 1 and 2", default=2)
	parser.add_argument("-mr", "--min-rating", type=int, default=30,
		help="specify the minimum accepted rating, default = 30")
	parser.add_argument("filename", help="output file, only supports FULL path (no tilde, etc.)")
	args = parser.parse_args()

	FILE_PATH = args.filename
	MIN_RATING_COUNT = args.min_rating

	try:
		if args.verbosity in (2, 3):
			start_time = time.time()
			with open(FILE_PATH, 'r') as file:
				init_filelength = sum(1 for i in file)

		while True:
			soup = get_html_source()
			book_rating_count = get_book_rating_count(soup)

			if to_int(book_rating_count) >= MIN_RATING_COUNT:
				try:
					book_pages = get_book_pages(soup)
					book_title = format_title(get_book_title(soup))
					book_rating = get_book_rating(soup)
					""" Re-check if file exists to catch deletion while program is running """
					if book_pages != "0" and os.path.isfile(FILE_PATH):
						with open(FILE_PATH, 'a') as file:
							file.write(book_title + ','  + book_pages + ',' + book_rating + '\n')

						if args.verbosity in (1, 3):
							sys.stdout.write("\033[F")
							with open(FILE_PATH, 'r') as file:
								print("\nLine count:", sum(1 for i in file), end='')
					elif not os.path.isfile(FILE_PATH):
						raise FileNotFoundError

				except AttributeError:
					continue

	except FileNotFoundError:
		print("File '", FILE_PATH, "' does not exist, or has been deleted.", sep='')

	except KeyboardInterrupt:
		if args.verbosity in (2, 3):
			with open(FILE_PATH, 'r') as file:
				lines_added = sum(1 for i in file) - init_filelength
			print("\nLines added: ", lines_added, " (since program start)", sep='')
			print("Run time: ", round(time.time() - start_time, 3), " seconds (real)", sep='')

if __name__ == '__main__':
	try:
		main()

	except ConnectionResetError:
		print("\nConnection reset by peer, exiting program.")
		pass
